rcl/test/rcl/test_publisher.cpp

```diff
@@ -372,38 +372,69 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_loan)
   }
 }
 
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_option) {
+  {
+    rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+    EXPECT_FALSE(publisher_options.disable_loaned_message);
+  }
+  {
+    ASSERT_TRUE(rcutils_set_env("ROS_DISABLE_LOANED_MESSAGES", "0"));
+    rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+    EXPECT_FALSE(publisher_options.disable_loaned_message);
+  }
+  {
+    ASSERT_TRUE(rcutils_set_env("ROS_DISABLE_LOANED_MESSAGES", "1"));
+    rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+    EXPECT_TRUE(publisher_options.disable_loaned_message);
+  }
+  {
+    ASSERT_TRUE(rcutils_set_env("ROS_DISABLE_LOANED_MESSAGES", "2"));
+    rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+    EXPECT_FALSE(publisher_options.disable_loaned_message);
+  }
+  {
+    ASSERT_TRUE(rcutils_set_env("ROS_DISABLE_LOANED_MESSAGES", "Unexpected"));
+    rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+    EXPECT_FALSE(publisher_options.disable_loaned_message);
+  }
+}
+
 TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_loan_disable) {
-  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
   const rosidl_message_type_support_t * ts =
     ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
   constexpr char topic_name[] = "pod_msg";
-  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
-  rcl_ret_t ret =
-    rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
-  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
-  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+
   {
-    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
-    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
-  });
+    bool is_fastdds = (std::string(rmw_get_implementation_identifier()).find("rmw_fastrtps") == 0);
+    rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+    rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+    rcl_ret_t ret =
+      rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    });
+    if (is_fastdds) {
+      EXPECT_TRUE(rcl_publisher_can_loan_messages(&publisher));
+    } else {
+      EXPECT_FALSE(rcl_publisher_can_loan_messages(&publisher));
+    }
```

Comment:
@fujitatomoya Could you please explain why there are different behavior for `rmw_fastrtps`?  
I think we shouldn't rely on default value of the environment variable in tests since it tend to be flaky or false negative/positive.
Instead we need explicitly set `publisher_options.disable_loaned_message` in code before publisher init to verify that our explicit options settings will override environment variable settings.

Reply:
@MichaelOrlov rmw_fastrtps just uses shared memory if possible as long as data type is POD, that is the default behavior for rmw_fastrtps, and this default behavior depends on rmw implementation. this default behavior verification was there before this PR, and i think we can keep those as well.

> explicitly set publisher_options.disable_loaned_message in code before publisher init to verify that our explicit options settings will override environment variable settings.

makes sense, i will add more test accordingly.

Reply:
@fujitatomoya I had in mind that if you will explicitly set `publisher_options.disable_loaned_message = true` before calling publisher init the further call for the `rcl_publisher_can_loan_messages(&publisher)` should return always false - regardless of the internal `rmw_fastrtps` implementation.
This is actually what we need to check.

Reply:
addressed by https://github.com/ros2/rcl/pull/1029/commits/c6f9fc3e78c8afdf5fd2f1ad274fb7825e05f36f